add flexcan support on imx8mp Cortex-A Core

JiafeiPan opened the pull request:
1. Update flexcan driver to support MMIO memory mapping.
2. Update CCM clock driver to add flexcan clock support.
3. Update flexcan dts binding to add rdc property.
4. Update SoC dts and board dts to add flexcan device tree nodes, add flexcan overlay dts, updated board document.
5. Update tests/drivers/can/api/testcase.yaml to add imx8mp flexcan support

JiafeiPan committed:
drivers: clock_control: mcux_ccm: add flexcan clock support

Enable FlexCAN clock support on imx8mp.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>

JiafeiPan committed:
dts: bindings: flexcan: add rdc property

Add rdc property to let RDC driver to configure RDC for flexcan.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>

JiafeiPan committed:
boards: imx8mp_evk: a53: enable FlexCAN on the board

Added dts nodes of FlexCAN1 and FlexCAN2 in SoC dts.
Added dts nodes and overlay in imx8mp_evk board.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>

JiafeiPan committed:
drivers: can: mcux: flexcan: add mmio mapping support

Use MMIO for device memory mapping, so that the driver can be used
both on MCU and MPU.

Add removed static MMU mapping in some platform accordingly.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>

JiafeiPan commented:
Hi, @henrikbrixandersen , would you please help a review on this PR? thanks.

JiafeiPan commented:
tests/drivers/can/api/ passes with the following log:

```

*** Booting Zephyr OS build v4.1.0-860-ga8e0bc31b7d2 ***
Running TESTSUITE can_classic
===================================================================
START - test_add_filter
 PASS - test_add_filter in 0.001 seconds
===================================================================
START - test_add_filter_without_callback
 PASS - test_add_filter_without_callback in 0.001 seconds
===================================================================
START - test_add_invalid_ext_filter
E: invalid filter with extended (29-bit) CAN ID 0x1fffffff, CAN ID mask 0x20000000
E: invalid filter with extended (29-bit) CAN ID 0x20000000, CAN ID mask 0x1fffffff
 PASS - test_add_invalid_ext_filter in 0.015 seconds
===================================================================
START - test_add_invalid_null_filter
 PASS - test_add_invalid_null_filter in 0.001 seconds
===================================================================
START - test_add_invalid_std_filter
E: invalid filter with standard (11-bit) CAN ID 0x7ff, CAN ID mask 0x800
E: invalid filter with standard (11-bit) CAN ID 0x800, CAN ID mask 0x7ff
 PASS - test_add_invalid_std_filter in 0.013 seconds
===================================================================
START - test_bitrate_limits
 PASS - test_bitrate_limits in 0.001 seconds
===================================================================
START - test_classic_get_capabilities
 PASS - test_classic_get_capabilities in 0.001 seconds
===================================================================
START - test_filters_added_while_stopped
 PASS - test_filters_added_while_stopped in 0.002 seconds
===================================================================
START - test_filters_preserved_through_bitrate_change
 PASS - test_filters_preserved_through_bitrate_change in 0.003 seconds
===================================================================
START - test_filters_preserved_through_mode_change
 PASS - test_filters_preserved_through_mode_change in 0.003 seconds
===================================================================
START - test_get_core_clock
 PASS - test_get_core_clock in 0.001 seconds
===================================================================
START - test_get_state
 PASS - test_get_state in 0.001 seconds
===================================================================
START - test_invalid_sample_point
 PASS - test_invalid_sample_point in 0.001 seconds
===================================================================
START - test_max_ext_filters
 PASS - test_max_ext_filters in 0.001 seconds
===================================================================
START - test_max_std_filters
 PASS - test_max_std_filters in 0.001 seconds
===================================================================
START - test_receive_timeout
 PASS - test_receive_timeout in 0.101 seconds
===================================================================
START - test_recover
 PASS - test_recover in 0.001 seconds
===================================================================
START - test_recover_while_stopped
 PASS - test_recover_while_stopped in 0.001 seconds
===================================================================
START - test_reject_ext_id_rtr
 PASS - test_reject_ext_id_rtr in 0.101 seconds
===================================================================
START - test_reject_std_id_rtr
 PASS - test_reject_std_id_rtr in 0.101 seconds
===================================================================
START - test_send_and_forget
 PASS - test_send_and_forget in 0.001 seconds
===================================================================
START - test_send_callback
 PASS - test_send_callback in 0.001 seconds
===================================================================
START - test_send_ext_id_dlc_of_range
E: DLC of 9 exceeds maximum (8)
 PASS - test_send_ext_id_dlc_of_range in 0.003 seconds
===================================================================
START - test_send_ext_id_out_of_range
E: invalid frame with extended (29-bit) CAN ID 0x20000000
 PASS - test_send_ext_id_out_of_range in 0.006 seconds
===================================================================
START - test_send_fd_format
E: unsupported CAN frame flags 0x04
 PASS - test_send_fd_format in 0.004 seconds
===================================================================
START - test_send_invalid_dlc
E: DLC of 9 exceeds maximum (8)
 PASS - test_send_invalid_dlc in 0.003 seconds
===================================================================
START - test_send_null_frame
 PASS - test_send_null_frame in 0.001 seconds
===================================================================
START - test_send_receive_ext_id
 PASS - test_send_receive_ext_id in 0.004 seconds
===================================================================
START - test_send_receive_ext_id_masked
 PASS - test_send_receive_ext_id_masked in 0.004 seconds
===================================================================
START - test_send_receive_ext_id_rtr
 SKIP - test_send_receive_ext_id_rtr in 0.001 seconds
===================================================================
START - test_send_receive_msgq
 PASS - test_send_receive_msgq in 0.010 seconds
===================================================================
START - test_send_receive_std_id
 PASS - test_send_receive_std_id in 0.004 seconds
===================================================================
START - test_send_receive_std_id_masked
 PASS - test_send_receive_std_id_masked in 0.004 seconds
===================================================================
START - test_send_receive_std_id_no_data
 PASS - test_send_receive_std_id_no_data in 0.001 seconds
===================================================================
START - test_send_receive_std_id_rtr
 SKIP - test_send_receive_std_id_rtr in 0.001 seconds
===================================================================
START - test_send_receive_wrong_id
 PASS - test_send_receive_wrong_id in 0.102 seconds
===================================================================
START - test_send_std_id_dlc_of_range
E: DLC of 9 exceeds maximum (8)
 PASS - test_send_std_id_dlc_of_range in 0.003 seconds
===================================================================
START - test_send_std_id_out_of_range
E: invalid frame with standard (11-bit) CAN ID 0x800
 PASS - test_send_std_id_out_of_range in 0.005 seconds
===================================================================
START - test_send_while_stopped
 PASS - test_send_while_stopped in 0.001 seconds
===================================================================
START - test_set_bitrate
 PASS - test_set_bitrate in 0.001 seconds
===================================================================
START - test_set_bitrate_too_high
 PASS - test_set_bitrate_too_high in 0.001 seconds
===================================================================
START - test_set_bitrate_too_low
 SKIP - test_set_bitrate_too_low in 0.001 seconds
===================================================================
START - test_set_bitrate_while_started
 PASS - test_set_bitrate_while_started in 0.001 seconds
===================================================================
START - test_set_mode_while_started
 PASS - test_set_mode_while_started in 0.001 seconds
===================================================================
START - test_set_state_change_callback
 PASS - test_set_state_change_callback in 0.001 seconds
===================================================================
START - test_set_timing_max
 PASS - test_set_timing_max in 0.001 seconds
===================================================================
START - test_set_timing_min
 PASS - test_set_timing_min in 0.001 seconds
===================================================================
START - test_set_timing_while_started
 PASS - test_set_timing_while_started in 0.001 seconds
===================================================================
START - test_start_while_started
 PASS - test_start_while_started in 0.001 seconds
===================================================================
START - test_stop_while_stopped
 PASS - test_stop_while_stopped in 0.001 seconds
===================================================================
TESTSUITE can_classic succeeded
Running TESTSUITE can_stats
===================================================================
START - test_can_stats_accessors
 PASS - test_can_stats_accessors in 0.001 seconds
===================================================================
TESTSUITE can_stats succeeded
Running TESTSUITE can_transceiver
===================================================================
START - test_get_transceiver
 PASS - test_get_transceiver in 0.001 seconds
===================================================================
TESTSUITE can_transceiver succeeded
Running TESTSUITE can_utilities
===================================================================
START - test_can_bytes_to_dlc
 PASS - test_can_bytes_to_dlc in 0.001 seconds
===================================================================
START - test_can_dlc_to_bytes
 PASS - test_can_dlc_to_bytes in 0.001 seconds
===================================================================
START - test_can_frame_matches_filter
 PASS - test_can_frame_matches_filter in 0.001 seconds
===================================================================
TESTSUITE can_utilities succeeded
Running TESTSUITE canfd
===================================================================
START - test_canfd_get_capabilities
 PASS - test_canfd_get_capabilities in 0.001 seconds
===================================================================
START - test_filters_preserved_through_classic_to_fd_mode_change
 PASS - test_filters_preserved_through_classic_to_fd_mode_change in 0.004 seconds
===================================================================
START - test_filters_preserved_through_fd_to_classic_mode_change
 PASS - test_filters_preserved_through_fd_to_classic_mode_change in 0.004 seconds
===================================================================
START - test_invalid_sample_point
 PASS - test_invalid_sample_point in 0.001 seconds
===================================================================
START - test_send_fd_dlc_out_of_range
E: DLC of 16 exceeds maximum (15)
 PASS - test_send_fd_dlc_out_of_range in 0.004 seconds
===================================================================
START - test_send_fd_incorrect_esi
E: unsupported CAN frame flags 0x14
 PASS - test_send_fd_incorrect_esi in 0.004 seconds
===================================================================
START - test_send_receive_classic
 PASS - test_send_receive_classic in 0.004 seconds
===================================================================
START - test_send_receive_fd
 PASS - test_send_receive_fd in 0.003 seconds
===================================================================
START - test_send_receive_mixed
 PASS - test_send_receive_mixed in 0.003 seconds
===================================================================
START - test_set_bitrate_data_too_low
 SKIP - test_set_bitrate_data_too_low in 0.001 seconds
===================================================================
START - test_set_bitrate_data_while_started
 PASS - test_set_bitrate_data_while_started in 0.001 seconds
===================================================================
START - test_set_bitrate_too_high
 PASS - test_set_bitrate_too_high in 0.001 seconds
===================================================================
START - test_set_timing_data_max
 PASS - test_set_timing_data_max in 0.001 seconds
===================================================================
START - test_set_timing_data_min
 PASS - test_set_timing_data_min in 0.001 seconds
===================================================================
START - test_set_timing_data_while_started
 PASS - test_set_timing_data_while_started in 0.001 seconds
===================================================================
TESTSUITE canfd succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [can_classic]: pass = 47, fail = 0, skip = 3, total = 50 duration = 0.521 seconds
 - PASS - [can_classic.test_add_filter] duration = 0.001 seconds
 - PASS - [can_classic.test_add_filter_without_callback] duration = 0.001 seconds
 - PASS - [can_classic.test_add_invalid_ext_filter] duration = 0.015 seconds
 - PASS - [can_classic.test_add_invalid_null_filter] duration = 0.001 seconds
 - PASS - [can_classic.test_add_invalid_std_filter] duration = 0.013 seconds
 - PASS - [can_classic.test_bitrate_limits] duration = 0.001 seconds
 - PASS - [can_classic.test_classic_get_capabilities] duration = 0.001 seconds
 - PASS - [can_classic.test_filters_added_while_stopped] duration = 0.002 seconds
 - PASS - [can_classic.test_filters_preserved_through_bitrate_change] duration = 0.003 seconds
 - PASS - [can_classic.test_filters_preserved_through_mode_change] duration = 0.003 seconds
 - PASS - [can_classic.test_get_core_clock] duration = 0.001 seconds
 - PASS - [can_classic.test_get_state] duration = 0.001 seconds
 - PASS - [can_classic.test_invalid_sample_point] duration = 0.001 seconds
 - PASS - [can_classic.test_max_ext_filters] duration = 0.001 seconds
 - PASS - [can_classic.test_max_std_filters] duration = 0.001 seconds
 - PASS - [can_classic.test_receive_timeout] duration = 0.101 seconds
 - PASS - [can_classic.test_recover] duration = 0.001 seconds
 - PASS - [can_classic.test_recover_while_stopped] duration = 0.001 seconds
 - PASS - [can_classic.test_reject_ext_id_rtr] duration = 0.101 seconds
 - PASS - [can_classic.test_reject_std_id_rtr] duration = 0.101 seconds
 - PASS - [can_classic.test_send_and_forget] duration = 0.001 seconds
 - PASS - [can_classic.test_send_callback] duration = 0.001 seconds
 - PASS - [can_classic.test_send_ext_id_dlc_of_range] duration = 0.003 seconds
 - PASS - [can_classic.test_send_ext_id_out_of_range] duration = 0.006 seconds
 - PASS - [can_classic.test_send_fd_format] duration = 0.004 seconds
 - PASS - [can_classic.test_send_invalid_dlc] duration = 0.003 seconds
 - PASS - [can_classic.test_send_null_frame] duration = 0.001 seconds
 - PASS - [can_classic.test_send_receive_ext_id] duration = 0.004 seconds
 - PASS - [can_classic.test_send_receive_ext_id_masked] duration = 0.004 seconds
 - SKIP - [can_classic.test_send_receive_ext_id_rtr] duration = 0.001 seconds
 - PASS - [can_classic.test_send_receive_msgq] duration = 0.010 seconds
 - PASS - [can_classic.test_send_receive_std_id] duration = 0.004 seconds
 - PASS - [can_classic.test_send_receive_std_id_masked] duration = 0.004 seconds
 - PASS - [can_classic.test_send_receive_std_id_no_data] duration = 0.001 seconds
 - SKIP - [can_classic.test_send_receive_std_id_rtr] duration = 0.001 seconds
 - PASS - [can_classic.test_send_receive_wrong_id] duration = 0.102 seconds
 - PASS - [can_classic.test_send_std_id_dlc_of_range] duration = 0.003 seconds
 - PASS - [can_classic.test_send_std_id_out_of_range] duration = 0.005 seconds
 - PASS - [can_classic.test_send_while_stopped] duration = 0.001 seconds
 - PASS - [can_classic.test_set_bitrate] duration = 0.001 seconds
 - PASS - [can_classic.test_set_bitrate_too_high] duration = 0.001 seconds
 - SKIP - [can_classic.test_set_bitrate_too_low] duration = 0.001 seconds
 - PASS - [can_classic.test_set_bitrate_while_started] duration = 0.001 seconds
 - PASS - [can_classic.test_set_mode_while_started] duration = 0.001 seconds
 - PASS - [can_classic.test_set_state_change_callback] duration = 0.001 seconds
 - PASS - [can_classic.test_set_timing_max] duration = 0.001 seconds
 - PASS - [can_classic.test_set_timing_min] duration = 0.001 seconds
 - PASS - [can_classic.test_set_timing_while_started] duration = 0.001 seconds
 - PASS - [can_classic.test_start_while_started] duration = 0.001 seconds
 - PASS - [can_classic.test_stop_while_stopped] duration = 0.001 seconds

SUITE PASS - 100.00% [can_stats]: pass = 1, fail = 0, skip = 0, total = 1 duration = 0.001 seconds
 - PASS - [can_stats.test_can_stats_accessors] duration = 0.001 seconds

SUITE PASS - 100.00% [can_transceiver]: pass = 1, fail = 0, skip = 0, total = 1 duration = 0.001 seconds
 - PASS - [can_transceiver.test_get_transceiver] duration = 0.001 seconds

SUITE PASS - 100.00% [can_utilities]: pass = 3, fail = 0, skip = 0, total = 3 duration = 0.003 seconds
 - PASS - [can_utilities.test_can_bytes_to_dlc] duration = 0.001 seconds
 - PASS - [can_utilities.test_can_dlc_to_bytes] duration = 0.001 seconds
 - PASS - [can_utilities.test_can_frame_matches_filter] duration = 0.001 seconds

SUITE PASS - 100.00% [canfd]: pass = 14, fail = 0, skip = 1, total = 15 duration = 0.034 seconds
 - PASS - [canfd.test_canfd_get_capabilities] duration = 0.001 seconds
 - PASS - [canfd.test_filters_preserved_through_classic_to_fd_mode_change] duration = 0.004 seconds
 - PASS - [canfd.test_filters_preserved_through_fd_to_classic_mode_change] duration = 0.004 seconds
 - PASS - [canfd.test_invalid_sample_point] duration = 0.001 seconds
 - PASS - [canfd.test_send_fd_dlc_out_of_range] duration = 0.004 seconds
 - PASS - [canfd.test_send_fd_incorrect_esi] duration = 0.004 seconds
 - PASS - [canfd.test_send_receive_classic] duration = 0.004 seconds
 - PASS - [canfd.test_send_receive_fd] duration = 0.003 seconds
 - PASS - [canfd.test_send_receive_mixed] duration = 0.003 seconds
 - SKIP - [canfd.test_set_bitrate_data_too_low] duration = 0.001 seconds
 - PASS - [canfd.test_set_bitrate_data_while_started] duration = 0.001 seconds
 - PASS - [canfd.test_set_bitrate_too_high] duration = 0.001 seconds
 - PASS - [canfd.test_set_timing_data_max] duration = 0.001 seconds
 - PASS - [canfd.test_set_timing_data_min] duration = 0.001 seconds
 - PASS - [canfd.test_set_timing_data_while_started] duration = 0.001 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL

```

JiafeiPan commented:
Rebased.

henrikbrixandersen reviewed (changes requested):
Very nice work, thanks. One question:

JiafeiPan commented on drivers/can/can_mcux_flexcan.c:
```diff
@@ -119,6 +128,11 @@ struct mcux_flexcan_data {
 #endif /* CONFIG_CAN_MCUX_FLEXCAN_FD */
 };
 
+static CAN_Type *get_base(const struct device *dev)
```
Sorry for late response, have updated, thanks.

JiafeiPan replied:
@henrikbrixandersen , would you please review this PR again? thanks.

henrikbrixandersen commented on boards/nxp/imx8mp_evk/imx8mp_evk_mimx8ml8_a53.yaml:
```diff
@@ -13,10 +13,11 @@ toolchain:
 supported:
-  - uart
-  - net
+  - can
```
Both `flexcan1` and `flexcan2` are disabled by default, so listing `can` as supported is incorrect.

Could one of the CAN controllers + transceiver be enabled by default? If yes, please remove the added test configuration in `tests/drivers/can/api`.

JiafeiPan replied:
I would like to remove can from supported list but not enable it by default, thanks.

JiafeiPan replied:
I think we need to enable device per requirement, for example, we only just want to run a hello world, why we need to enable flexcan in hello world application? considering the image size and running performance, the application is as good as smaller. I agree with you that it is easy for customer to run any usecase if we enable all the device by default,  so as you said if this is normally done in Zephyr, I would like to align with the other platforms.

JiafeiPan replied:
Updated, please review again, thanks @henrikbrixandersen

alexanderwachter replied:
Brix is the Maintainer of CAN bus.
Nothing can related will go in without his approval.

JiafeiPan commented on soc/nxp/imx/imx9/imx93/a55/mmu_regions.c:
```diff
@@ -32,9 +32,6 @@ static const struct arm_mmu_region mmu_regions[] = {
-	MMU_REGION_DT_COMPAT_FOREACH_FLAT_ENTRY(nxp_flexcan,
-						(MT_DEVICE_nGnRnE | MT_P_RW_U_NA | MT_NS))
-
 #if CONFIG_SOF
```
yes, indeed, have merge into one commit, thanks.

Zhiqiang-Hou commented on drivers/can/can_mcux_flexcan.c:
```diff
@@ -389,7 +405,10 @@ static int mcux_flexcan_stop(const struct device *dev)
 static int mcux_flexcan_set_mode(const struct device *dev, can_mode_t mode)
 {
 	can_mode_t supported = CAN_MODE_LOOPBACK | CAN_MODE_LISTENONLY | CAN_MODE_3_SAMPLES;
+#ifdef CONFIG_CAN_MCUX_FLEXCAN_FD
```
I see that there's still code needs the `config` in this condiation: 
`if (UTIL_AND(IS_ENABLED(CONFIG_CAN_MCUX_FLEXCAN_FD), config->flexcan_fd))`

JiafeiPan replied:
yes.

Zhiqiang-Hou reviewed (dismissed):
Looks good but a nit on comment.

JiafeiPan commented on boards/nxp/imx8mp_evk/Kconfig.defconfig:
```diff
@@ -39,6 +39,15 @@ endif # NET_SOCKETS_SERVICE
+config CAN_TRANSCEIVER_INIT_PRIORITY
+	default 75
+
+endif # CAN
```
fixed, thanks.

JiafeiPan commented:
Rebased and address @Zhiqiang-Hou 's comments.

henrikbrixandersen reviewed (changes requested):
You just pushed new changes within the last 24 hours. Please allow me to review these.

JiafeiPan commented:
> You just pushed new changes within the last 24 hours. Please allow me to review these.

Hi, @henrikbrixandersen, I have submitted this PR for four months, and addressed your changes request for more than one month, not just 24 hours,  so I removed your changes request as I don't get your feedback so long a time, I hope it could be merged in zephyr v4.2, please let me know how we could merge it ASAP, thanks.

JiafeiPan commented on boards/nxp/imx8mp_evk/imx8mp_evk_mimx8ml8_a53.dts:
```diff
@@ -38,6 +39,22 @@
+		max-bitrate = <8000000>;
+		#phy-cells = <0>;
+		status = "disabled";
+	};
```
Both PHY1 and PHY2 has enable GPIO, PHY2 is disabled by default, but PHY1 is enabled by default configuration, anyway I also added enable-gpio for PHY1 to make sure it is enabled for any case, thanks.

JiafeiPan commented:
Hi, @henrikbrixandersen , have addressed your comments, please review again, thanks.

henrikbrixandersen reviewed (approved):
Thanks!
